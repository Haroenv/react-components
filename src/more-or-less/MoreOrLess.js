import React from 'react';
import { pure, props, t, skinnable } from 'revenge';
import { Icon } from 'Basic';

import './moreOrLess.scss';

@pure
@skinnable()
@props({
  children: t.ReactNode,
  className: t.maybe(t.Str),
  expanded: t.Bool,
  onExpandedChange: t.Func
})
export default class MoreOrLess extends React.Component {

  toggleExpanded = () => {
    this.props.onExpandedChange(!this.props.expanded);
  }

  getLocals() {
    const panelState = this.props.expanded ? 'more' : 'less';
    const className = `more-or-less ${panelState} ${this.props.className}`;
    const { toggleExpanded } = this;
    const icon = this.props.expanded ? 'angle-up' : 'angle-down';
    return {
      content: this.props.children,
      className,
      icon,
      toggleExpanded
    };
  }

  templateExpandButton = ({ icon, toggleExpanded }) => {
    return (
      <div className="expand-button" onClick={toggleExpanded}>
        <Icon icon={icon} className="expand-button-icon" />
      </div>
    );
  }

  template({ className, content, icon, toggleExpanded }) {
    return (
      <div className={className}>
        {content}
        {this.templateExpandButton({ icon, toggleExpanded })}
      </div>
    );
  }
}
