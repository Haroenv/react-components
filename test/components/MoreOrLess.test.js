import React from 'react';
import render from '../render';

import MoreOrLess from '../../src/more-or-less';

const exampleProps = {
  children: 'content',
  className: 'fancy-class-name',
  icons: {
    expanded: 'angle-up',
    collapsed: 'angle-down'
  },
  onExpandedChange: () => {}
};

const componentMore = new MoreOrLess({
  ...exampleProps,
  expanded: true
});

const componentLess = new MoreOrLess({
  ...exampleProps,
  expanded: false
});


describe('MoreOrLess', () => {

  it('renders correctly when expanded', () => {
    const tree = render(
      <MoreOrLess {...exampleProps} expanded />
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when not expanded', () => {
    const tree = render(
      <MoreOrLess {...exampleProps} expanded={false} />
    );
    expect(tree).toMatchSnapshot();
  });

  describe('getLocals', () => {

    it('computes className', () => {
      const { wrapperProps: { className: classNameMore } } = componentMore.getLocals();
      expect(classNameMore).toBe('more-or-less more fancy-class-name');

      const { wrapperProps: { className: classNameLess } } = componentLess.getLocals();
      expect(classNameLess).toBe('more-or-less less fancy-class-name');
    });

    it('computes icon', ()  => {
      const { icon: iconMore } = componentMore.getLocals();
      expect(iconMore).toBe('angle-up');

      const { icon: iconLess } = componentLess.getLocals();
      expect(iconLess).toBe('angle-down');
    });

  });

});

