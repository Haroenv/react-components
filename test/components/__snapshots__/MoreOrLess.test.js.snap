exports[`MoreOrLess renders correctly when expanded 1`] = `
"<FlexView column={true} shrink={false} style={[undefined]} className=\"more-or-less more fancy-class-name\">
content
<FlexView hAlignContent=\"center\" vAlignContent=\"center\" className=\"expand-button\" onClick={[Function]} shrink={false}>
<Icon icon=\"angle-up\" className=\"expand-button-icon\" paths={1} onClick={[Function]} />
</FlexView>
</FlexView>"
`;

exports[`MoreOrLess renders correctly when not expanded 1`] = `
"<FlexView column={true} shrink={false} style={[undefined]} className=\"more-or-less less fancy-class-name\">
content
<FlexView hAlignContent=\"center\" vAlignContent=\"center\" className=\"expand-button\" onClick={[Function]} shrink={false}>
<Icon icon=\"angle-down\" className=\"expand-button-icon\" paths={1} onClick={[Function]} />
</FlexView>
</FlexView>"
`;
